File: Networking/src/main/java/protocolBuffers/ProtocolBufferWorker.java
package protocolBuffers;

import DTO.EventDTO;
import DTO.ParticipantDTO;
import contestUtils.IContestServices;
import contestUtils.IMainObserver;
import example.example.Event;
import example.example.Office;
import example.example.Participant;
import example.example.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.List;

public class ProtocolBufferWorker implements Runnable, IMainObserver {
    private final IContestServices contestServices;
    private final Socket socket;
    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    private final static Logger logger = LogManager.getLogger();

    public ProtocolBufferWorker(IContestServices contestServices, Socket socket) {
        this.contestServices = contestServices;
        this.socket = socket;
        try {
            input = socket.getInputStream();
            output = socket.getOutputStream();
            connected = true;
        } catch (Exception e) {
            logger.error("Error initializing worker: " + e.getMessage());
        }
    }

    @Override
    public void run() {
        while (connected) {
            try {
                SwimmingContestProtocol.SwimmingContestRequest request =
                        SwimmingContestProtocol.SwimmingContestRequest.parseDelimitedFrom(input);

                if (request != null) {
                    SwimmingContestProtocol.SwimmingContestResponse response = handleRequest(request);
                    if (response != null) {
                        sendResponse(response);
                    }
                }

                Thread.sleep(100);
            } catch (Exception e) {
                logger.error("Error in worker loop: " + e.getMessage(), e);
                connected = false;
            }
        }

        try {
            input.close();
            output.close();
            socket.close();
        } catch (Exception e) {
            logger.error("Error closing connection: " + e.getMessage(), e);
        }
    }

    private void sendResponse(SwimmingContestProtocol.SwimmingContestResponse response) throws Exception {
        response.writeDelimitedTo(output);
        output.flush();
    }

    private SwimmingContestProtocol.SwimmingContestResponse handleRequest(SwimmingContestProtocol.SwimmingContestRequest request) {
        switch (request.getType()) {
            case Login -> {
                try {
                    logger.info("Login attempt for user: " + request.getUserName());
                    User user = contestServices.login(request.getUserName(), request.getPassword(), this);
                    logger.info("Login successful for user: " + user.getUserName());
                    return ProtoBuilderUtils.createOkResponse(user);
                } catch (Exception e) {
                    logger.error("Login failed: " + e.getMessage());
                    return ProtoBuilderUtils.createErrorResponse(e.getMessage());
                }
            }

            case Logout -> {
                System.out.println("> handling logout request");
                final User user = ProtoBuilderUtils.getUser(request);

                try {
                    contestServices.logout(user, this);
                    connected = false;
                    return ProtoBuilderUtils.createOkResponse();
                } catch (Exception contestDataException) {
                    return ProtoBuilderUtils.createErrorResponse(contestDataException.getMessage());
                }
            }

            case CreateParticipant -> {
                try {
                    Participant participant = ProtoBuilderUtils.getParticipant(request);
                    contestServices.saveParticipant(participant);
                    return ProtoBuilderUtils.createNewParticipantResponse(participant);
                } catch (Exception e) {
                    return ProtoBuilderUtils.createErrorResponse(e.getMessage());
                }
            }

            case CreateEventEntry -> {
                try {
                    List<Office> offices = ProtoBuilderUtils.getEventEntries(request);
                    contestServices.saveEventEntry(offices);
                    return ProtoBuilderUtils.createUpdatedEventsResponse(contestServices.getEventsWithParticipantsCount().stream().toList());
                } catch (Exception e) {
                    return ProtoBuilderUtils.createErrorResponse(e.getMessage());
                }
            }

            case GetEventsWithParticipantsCount -> {
                try {
                    List<EventDTO> events = contestServices.getEventsWithParticipantsCount().stream().toList();
                    return ProtoBuilderUtils.createEventsWithParticipantsCountResponse(events);
                } catch (Exception e) {
                    return ProtoBuilderUtils.createErrorResponse(e.getMessage());
                }
            }

            case GetAllParticipants -> {
                try {
                    List<Participant> participants = contestServices.findAllParticipants().stream().toList();
                    return ProtoBuilderUtils.createAllParticipantsResponse(participants);
                } catch (Exception e) {
                    return ProtoBuilderUtils.createErrorResponse(e.getMessage());
                }
            }

            case GetAllEvents -> {
                try {
                    List<Event> events = contestServices.findAllEvents().stream().toList();
                    return ProtoBuilderUtils.createAllEventsResponse(events);
                } catch (Exception e) {
                    return ProtoBuilderUtils.createErrorResponse(e.getMessage());
                }
            }

            case GetParticipantsForEventWithCount -> {
                try {
                    long eventId = request.getEventId();
                    List<ParticipantDTO> participants = contestServices.getParticipantsForEventWithCount(eventId).stream().toList();
                    return ProtoBuilderUtils.createParticipantsForEventWithCountResponse(participants);
                } catch (Exception e) {
                    return ProtoBuilderUtils.createErrorResponse(e.getMessage());
                }
            }

            default -> {
                return ProtoBuilderUtils.createErrorResponse("Unknown request type: " + request.getType());
            }
        }
    }

    @Override
    public void participantAdded(Participant participant) {
        try {
            sendResponse(ProtoBuilderUtils.createNewParticipantResponse(participant));
        } catch (Exception e) {
            logger.error("Failed to notify about participant: " + e.getMessage(), e);
        }
    }

    @Override
    public void newRegistration(List<EventDTO> events) {
        try {
            sendResponse(ProtoBuilderUtils.createUpdatedEventsResponse(events));
        } catch (Exception e) {
            logger.error("Failed to notify about new registration: " + e.getMessage(), e);
        }
    }

    @Override
    public void eventAdded(EventDTO event) {
        try {
            sendResponse(ProtoBuilderUtils.createNewEventEntryResponse(event));
        } catch (Exception e) {
            logger.error("Failed to notify about new event: " + e.getMessage(), e);
        }
    }
}

File: java-app/csharp-app/Networking/ProtocolBuffers/ProtocolBufferWorker.cs
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using Google.Protobuf;
using log4net;
using Org.Example.Protocolbuffers;
using Service;

using Participant = mpp_proiect_csharp_DianaGliga11.Model.Participant;
using User = mpp_proiect_csharp_DianaGliga11.Model.User;
using Event = mpp_proiect_csharp_DianaGliga11.Model.Event;
using EventDTO = mpp_proiect_csharp_DianaGliga11.Model.DTO.EventDTO;
using ParticipantDTO = mpp_proiect_csharp_DianaGliga11.Model.DTO.ParticipantDTO;
using Office = mpp_proiect_csharp_DianaGliga11.Model.Office;

namespace Networking.ProtocolBuffers;

public class ProtocolBufferWorker : IMainObserver
{
    private readonly IContestServices _server;
    private readonly TcpClient _connection;
    private NetworkStream _stream;
    private volatile bool _connected;
    private static readonly ILog Log = LogManager.GetLogger(typeof(ClientWorker));

    public ProtocolBufferWorker(IContestServices server, TcpClient connection)
    {
        _server = server;
        _connection = connection;

        try
        {
            _stream = _connection.GetStream();
            _connected = true;
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }

    public virtual void Run()
    {
        while (_connected)
        {
            try
            {
                SwimmingContestRequest request = SwimmingContestRequest.Parser.ParseDelimitedFrom(_stream);
                SwimmingContestResponse response = HandleRequest(request);

                if (response != null)
                {
                    SendResponse(response);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }

            try
            {
                Thread.Sleep(1000);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        try
        {
            _stream.Close();
            _connection.Close();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }

    private SwimmingContestResponse HandleRequest(SwimmingContestRequest request)
    {
        try
        {
            switch (request.Type)
            {
                case SwimmingContestRequest.Types.Type.Login:
                    Log.Debug("Login request...");
                    string username = ProtocolBuilderUtils.GetUsername(request);
                    string password = ProtocolBuilderUtils.GetPassword(request);
                    try
                    {
                       User user = _server.Login(username, password, this);
                       return ProtocolBuilderUtils.CreateOkResponse(user);
                    }
                    catch (Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case SwimmingContestRequest.Types.Type.Logout:
                    Log.Debug("Logout request...");
                    User deconncectedUser = ProtocolBuilderUtils.GetUser(request);
                    try
                    {
                        _server.Logout(deconncectedUser, this);
                        _connected = false;
                        return ProtocolBuilderUtils.CreateOkResponse(deconncectedUser);
                    }
                    catch (Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case SwimmingContestRequest.Types.Type.CreateParticipant:
                    Log.Debug("Create participant request...");
                    Participant participant = ProtocolBuilderUtils.GetParticipant(request);
                    try
                    {
                        _server.SaveParticipant(participant, this);
                        return ProtocolBuilderUtils.CreateNewParticipantResponse(participant);
                    }catch(Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case SwimmingContestRequest.Types.Type.GetAllParticipants:
                    Log.Debug("Get all participants request...");
                    try
                    {
                        List<Participant> participants = _server.GetAllParticipants();
                        return ProtocolBuilderUtils.CreateAllParticipantsResponse(participants);
                    }catch(Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case SwimmingContestRequest.Types.Type.GetAllEvents:
                    Log.Debug("Get all events request...");
                    try
                    {
                        List<Event> events = _server.GetAllEvents();
                        return ProtocolBuilderUtils.CreateAllEventsResponse(events);
                    }catch(Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case SwimmingContestRequest.Types.Type.GetEventsWithParticipantsCount:
                    Log.Debug("Get events with count request...");
                    try
                    {
                        List<EventDTO> events = _server.GetEventsWithParticipantsCount();
                        return ProtocolBuilderUtils.CreateEventsWithParticipantsCountResponse(events);
                    }catch(Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case SwimmingContestRequest.Types.Type.GetParticipantsForEventWithCount:
                    Log.Debug("Get participants for event request...");
                    try
                    {
                        long eventId = request.EventId;
                        List<ParticipantDTO> participants = _server.GetParticipantsForEventWithCount(eventId);
                        return ProtocolBuilderUtils.CreateParticipantsForEventWithCountResponse(participants);
                    }catch(Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                case SwimmingContestRequest.Types.Type.CreateEventEntry:
                    Log.Debug("Create event entries request...");
                    List<Office> eventEntries = ProtocolBuilderUtils.GetEventEntries(request);
                    try
                    {
                        _server.SaveEventsEntries(eventEntries);
                        return ProtocolBuilderUtils.CreateUpdatedEventsResponse(
                            _server.GetEventsWithParticipantsCount());
                    }catch(Exception e)
                    {
                        _connected = false;
                        return ProtocolBuilderUtils.CreateErrorResponse(e.Message);
                    }

                default:
                    throw new InvalidOperationException($"Unsupported request type: {request.Type}");
            }
        }
        catch (Exception ex)
        {
            Log.Error("HandleRequest error", ex);
            return ProtocolBuilderUtils.CreateErrorResponse(ex.Message);
        }
    }

    private void SendResponse(SwimmingContestResponse response)
    {
        lock (_stream)
        {
            response.WriteDelimitedTo(_stream);
            _stream.Flush();
            Log.Debug($"Sent response of type: {response.Type}");
        }
    }
    public void ParticipantAdded(Participant participant)
    {
        Log.Debug($"Observer: participant added {participant.Name}");
        SendResponse(ProtocolBuilderUtils.CreateNewParticipantResponse(participant));
    }

    public void EventEvntriesAdded(List<EventDTO> updatedEvents)
    {
        Log.Debug($"Observer: events updated ({updatedEvents.Count})");
        SendResponse(ProtocolBuilderUtils.CreateUpdatedEventsResponse(updatedEvents));
    }
    
}